File: app/csharp/AccountingSystem.Tests/Infrastructure/Repositories/TaxTransactionRepositoryTest.cs
using AccountingSystem.Domain.Entities;
using AccountingSystem.Infrastructure.Persistence.Repositories;
using FluentAssertions;
using Xunit;

namespace AccountingSystem.Tests.Infrastructure.Repositories;

/// <summary>
/// 課税取引マスタ - Dapper 統合テスト
/// </summary>
public class TaxTransactionRepositoryTest : DatabaseTestBase
{
    private TaxTransactionRepository CreateRepository() => new TaxTransactionRepository(ConnectionString);

    [Fact(DisplayName = "初期データが投入されている")]
    public async Task TestInitialData()
    {
        var repository = CreateRepository();

        var all = await repository.FindAllAsync();
        all.Should().HaveCount(4);

        var codes = all.Select(t => t.TaxCode).ToList();
        codes.Should().Contain(new[] { "01", "02", "03", "04" });
    }

    [Fact(DisplayName = "課税取引コードで検索できる")]
    public async Task TestFindByCode()
    {
        var repository = CreateRepository();

        var taxable = await repository.FindByCodeAsync("01");

        taxable.Should().NotBeNull();
        taxable!.TaxCode.Should().Be("01");
        taxable.TaxName.Should().Be("課税");
        taxable.TaxRate.Should().Be(0.10m);
    }

    [Fact(DisplayName = "有効な課税取引のみ取得できる")]
    public async Task TestFindActive()
    {
        var repository = CreateRepository();

        // すべて有効なので4件取得される
        var active = await repository.FindActiveAsync();
        active.Should().HaveCount(4);
    }

    [Fact(DisplayName = "新しい課税取引を登録できる")]
    public async Task TestInsert()
    {
        var repository = CreateRepository();

        var newTax = new TaxTransaction
        {
            TaxCode = "05",
            TaxName = "軽減税率",
            TaxRate = 0.08m,
            Description = "食料品等に適用される軽減税率",
            IsActive = true
        };

        await repository.InsertAsync(newTax);

        var found = await repository.FindByCodeAsync("05");
        found.Should().NotBeNull();
        found!.TaxName.Should().Be("軽減税率");
        found.TaxRate.Should().Be(0.08m);
    }

    [Fact(DisplayName = "課税取引を更新できる")]
    public async Task TestUpdate()
    {
        var repository = CreateRepository();

        // 既存の課税取引を取得
        var existing = await repository.FindByCodeAsync("01");
        existing.Should().NotBeNull();

        // 更新
        var updated = existing! with { TaxRate = 0.12m, Description = "更新後の説明" };
        await repository.UpdateAsync(updated);

        // 検証
        var found = await repository.FindByCodeAsync("01");
        found!.TaxRate.Should().Be(0.12m);
        found.Description.Should().Be("更新後の説明");
    }

    [Fact(DisplayName = "課税取引を削除できる")]
    public async Task TestDelete()
    {
        var repository = CreateRepository();

        // 新しいデータを登録してから削除
        var newTax = new TaxTransaction
        {
            TaxCode = "99",
            TaxName = "テスト用",
            TaxRate = 0.00m,
            IsActive = true
        };
        await repository.InsertAsync(newTax);

        await repository.DeleteAsync("99");

        var deleted = await repository.FindByCodeAsync("99");
        deleted.Should().BeNull();
    }
}

File: app/fsharp/AccountingSystem.Tests/Infrastructure/TaxTransactionRepositoryTest.fs
module AccountingSystem.Tests.Infrastructure.TaxTransactionRepositoryTest

open AccountingSystem.Domain.Models
open AccountingSystem.Infrastructure.Persistence.Repositories.TaxTransactionRepository
open AccountingSystem.Tests.DatabaseTestBase
open Xunit
open FsUnit.Xunit

/// <summary>
/// 課税取引マスタ - Dapper 統合テスト
/// </summary>
type TaxTransactionRepositoryTest() =
    inherit DatabaseTestBase()

    [<Fact>]
    member this.``初期データが投入されている``() =
        task {
            let! all = findAllAsync this.ConnectionString
            all |> List.length |> should be (greaterThanOrEqualTo 8)

            let! taxable = findByCodeAsync this.ConnectionString "01"
            taxable.IsSome |> should equal true
            taxable.Value.TaxName |> should equal "課税売上10%"
            taxable.Value.TaxRate |> should equal 10.00m
        }

    [<Fact>]
    member this.``課税取引コードで検索できる``() =
        task {
            let! found = findByCodeAsync this.ConnectionString "02"
            found.IsSome |> should equal true
            found.Value.TaxName |> should equal "課税売上8%（軽減）"
            found.Value.TaxRate |> should equal 8.00m
        }

    [<Fact>]
    member this.``動的SQL_課税取引名で部分一致検索``() =
        task {
            // "課税"を含む取引を検索
            let! results = searchAsync this.ConnectionString (Some "課税") None
            results |> List.length |> should be (greaterThanOrEqualTo 4)
            results |> List.forall (fun t -> t.TaxName.Contains("課税")) |> should equal true
        }

    [<Fact>]
    member this.``動的SQL_最小税率で検索``() =
        task {
            // 税率が5以上の取引を検索（10%, 8%）
            let! results = searchAsync this.ConnectionString None (Some 5.00m)
            results |> List.length |> should be (greaterThanOrEqualTo 4)
            results |> List.forall (fun t -> t.TaxRate >= 5.00m) |> should equal true
        }

    [<Fact>]
    member this.``存在しない課税取引コードはNoneを返す``() =
        task {
            let! found = findByCodeAsync this.ConnectionString "99"
            found.IsNone |> should equal true
        }
